src/Spout/Writer/WriterInterface.php

```diff
@@ -34,26 +34,38 @@ public function openToBrowser($outputFileName);
     /**
      * Write given data to the output. New data will be appended to end of stream.
      *
-     * @param  array $dataRow Array containing data to be streamed.
-     *          Example $dataRow = ['data1', 1234, null, '', 'data5'];
+     * @param array|\Box\Spout\Writer\Common\Entity\Row $row The row to be appended to the stream
      * @return WriterInterface
-     * @throws \Box\Spout\Writer\Exception\WriterNotOpenedException If the writer has not been opened yetthe writer
-     * @throws \Box\Spout\Common\Exception\IOException If unable to write data
+     * @internal param array $row Array containing data to be streamed.
+     *          Example $row= ['data1', 1234, null, '', 'data5'];
+     * @internal param \Box\Spout\Writer\Common\Entity\Row $row A Row object with cells and styles
+     *          Example $row = (new Row())->addCell('data1');
      */
-    public function addRow(array $dataRow);
+    public function addRow($row);
 
     /**
      * Write given data to the output and apply the given style.
      * @see addRow
      *
-     * @param array $dataRow Array of array containing data to be streamed.
+     * @param array|\Box\Spout\Writer\Common\Entity\Row $row The row to be appended to the stream
      * @param Style $style Style to be applied to the row.
      * @return WriterInterface
-     * @throws \Box\Spout\Common\Exception\InvalidArgumentException If the input param is not valid
-     * @throws \Box\Spout\Writer\Exception\WriterNotOpenedException If this function is called before opening the writer
-     * @throws \Box\Spout\Common\Exception\IOException If unable to write data
+     * @internal param array $row Array containing data to be streamed.
+     *          Example $row= ['data1', 1234, null, '', 'data5'];
+     * @internal param \Box\Spout\Writer\Common\Entity\Row $row A Row object with cells and styles
+     *          Example $row = (new Row())->addCell('data1');
+     */
+    public function addRowWithStyle($row, $style);
+
+    /**
+     * Write given data to the output with a closure funtion. New data will be appended to end of stream.
+     *
+     * @param \Closure $callback A callback returning a Row object. A new Row object is injected into the callback.
+     * @return WriterInterface
+     * @internal param \Closure $callback
+     *          Example withRow(function(Row $row) { return $row->addCell('data1'); })
      */
-    public function addRowWithStyle(array $dataRow, $style);
+    public function withRow(\Closure $callback);
```

Comment:
I'm still unsure if this function is a good idea... Having multiple ways to do something is not necessarily better